include/dsn/dist/replication/replication_app_client_base.h

```diff
@@ -346,13 +394,28 @@ namespace dsn { namespace replication {
             dsn_message_t         request;
 
             ::dsn::service::zlock lock; // [
-            dsn::task_ptr         timeout_timer; // when partition config is unknown at the first place            
+            dsn::task_ptr         timeout_timer; // when partition config is unknown at the first place
             bool                  completed;
             // ]
         };
-
         typedef ::dsn::ref_ptr<request_context> request_context_ptr;
 
+        struct meta_context : public ref_counter
+        {
+            dsn_message_t           request;
+            dsn::task_ptr           callback_task;
+            uint64_t                timeout_ts_us; // timeout at this timing point
+
+            dsn::service::zlock     lock; // [
+            task_ptr                timeout_timer; // when partition config is unknown at the first place
+            bool                    completed;
+            // ]
+        };
```

Comment:
I don't get it what this meta context is for. For replica request, because in certain cases, they need to be delayed until the target address (primary) is decided, where therefore need a replica context to maintain all these complexities. For meta request, the target address is a group address which is read from config file, therefore a simple rpc will work. (if the leader changes, the rdsn core will automatically retry with the new leader).